internal/services/kusto/kusto_cosmosdb_data_connection_resource.go

```diff
@@ -0,0 +1,248 @@
+		},
+		"retrieval_start_date": {
+			Type:         pluginsdk.TypeString,
+			Optional:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.StringIsNotEmpty,
+		},
+	}
+}
+
+func (r CosmosDBDataConnectionResource) Attributes() map[string]*schema.Schema {
+	return map[string]*schema.Schema{}
+}
+
+func (r CosmosDBDataConnectionResource) ModelObject() interface{} {
+	return &CosmosDBDataConnectionModel{}
+}
+
+func (r CosmosDBDataConnectionResource) ResourceType() string {
+	return "azurerm_kusto_cosmosdb_data_connection"
+}
+
+func (r CosmosDBDataConnectionResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+
+		// the Func returns a function which retrieves the current state of the Resource Group into the state
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var model CosmosDBDataConnectionModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding %s: %+v", r.ResourceType(), err)
+			}
+
+			client := metadata.Client.Kusto.DataConnectionsClient
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			id := dataconnections.NewDataConnectionID(subscriptionId, model.ResourceGroupName, model.ClusterName, model.DatabaseName, model.Name)
+
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for presence of existing %s: %+v", id, err)
+			}
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			properties := dataconnections.CosmosDbDataConnectionProperties{
+				CosmosDbAccountResourceId: model.CosmosDbAccountId,
+				CosmosDbContainer:         model.CosmosDbContainer,
+				CosmosDbDatabase:          model.CosmosDbDatabase,
+				TableName:                 model.TableName,
+				ManagedIdentityResourceId: model.ManagedIdentityId,
+			}
+
+			if model.MappingRuleName != "" {
+				properties.MappingRuleName = &model.MappingRuleName
+			}
+
+			if model.RetrievalStartDate != "" {
+				properties.RetrievalStartDate = &model.RetrievalStartDate
+			}
+
+			l := location.Normalize(model.Location)
+
+			dataConnection := dataconnections.CosmosDbDataConnection{
+				Location:   &l,
+				Name:       &model.Name,
+				Properties: &properties,
+			}
+
+			if err := client.CreateOrUpdateThenPoll(ctx, id, dataConnection); err != nil {
+				return fmt.Errorf("creating %s: %+v", id, err)
+			}
+
+			metadata.SetID(id)
+			return nil
+		},
+	}
+}
+
+func (r CosmosDBDataConnectionResource) Read() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Kusto.DataConnectionsClient
+
+			id, err := dataconnections.ParseDataConnectionID(metadata.ResourceData.Id())
+			if err != nil {
+				return fmt.Errorf("parsing %s: %+v", metadata.ResourceData.Id(), err)
```

Comment:
The parsers already return detailed error messages, we should just pass them along
```suggestion
				return err
```